fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,458 @@
-        elif self._write_function_type is FunctionType.VECTOR:
-            if self._can_apply_2d_3d_coupling():
-                # in 2d-3d coupling z dimension is set to zero
-                precice_write_data = np.column_stack((self._write_data[:, 0],
-                                                      self._write_data[:, 1],
-                                                      np.zeros(self._n_vertices)))
+        write_data = convert_fenics_to_precice(write_function, self._coupling_mesh_vertices)
 
-                assert(precice_write_data.shape[0] == self._n_vertices and
-                       precice_write_data.shape[1] == self._dimensions)
+        write_data_id = self._interface.get_data_id(self._config.get_write_data_name(),
+                                                    self._interface.get_mesh_id(self._config.get_coupling_mesh_name()))
 
-                self._interface.write_block_vector_data(self._write_data_id, self._vertex_ids, precice_write_data)
-                    
-            elif self._fenics_dimensions == self._dimensions:
-                self._interface.write_block_vector_data(self._write_data_id, self._vertex_ids, self._write_data)
+        if write_function_type is FunctionType.SCALAR:
+            self._interface.write_block_scalar_data(write_data_id, self._vertex_ids, write_data)
+        elif write_function_type is FunctionType.VECTOR:
+            if self._apply_2d_3d_coupling:
+                # in 2d-3d coupling z dimension is set to zero
+                precice_write_data = np.column_stack((write_data[:, 0], write_data[:, 1], np.zeros(self._n_vertices)))
+                assert (precice_write_data.shape[0] == self._n_vertices and
+                        precice_write_data.shape[1] == self._interface.get_dimensions())
+                self._interface.write_block_vector_data(write_data_id, self._vertex_ids, precice_write_data)
+            elif self._fenics_dimensions == self._interface.get_dimensions():
+                self._interface.write_block_vector_data(write_data_id, self._vertex_ids, write_data)
             else:
                 raise Exception("Dimensions don't match.")
         else:
             raise Exception("Rank of function space is neither 0 nor 1")
 
-    def _read_block_data(self):
-        """ Reads data from preCICE. Depending on the dimensions of the simulation (2D-3D Coupling, 2D-2D coupling or
-        Scalar/Vector write function) read_data is converted.
-
-        Note: For quasi 2D fenics in a 3D coupled simulation the y component of the vectors is deleted.
+    def initialize(self, coupling_subdomain, mesh, read_function, function_space, dimensions=2):
+        """
+        Initializes the coupling interface and sets up the mesh in preCICE.
+
+        Parameters
+        ----------
+        coupling_subdomain : FEniCS Domain
+            Part of FEniCS Mesh which is the coupling interface.
```

Comment:
I assume that the "coupling interface" here is not the same thing like the "coupling interface" five lines above. We should make sure to be exact about wording. Here, the term interface is already occupied by the interface from computer science.

I would suggest to use `coupling_boundary` instead, since we usually talk about a "coupling mesh" (see https://github.com/precice/precice/wiki/Adapter-Example#mesh-and-data-access). However, in the FEniCS adapter we do not require a discrete mesh, but a (continuous) subdomain.